ingest-geoip: establish boundaries

yaauie opened the pull request:
This large-scale refactor is net zero change to the *usage* of the geoip processor, but simplifies the public `GeoIpDatabase` interface to reduce the complexity of implementing it from outside of the Elasticsearch codebase, and to make implmentations less fragile as new database types are added to this codebase.

It also partially opens the door to the support of non-Maxmind enrichments, or at least provided a prototype that we can learn from.

_This PR is meant to be a starting point for a conversation._

yaauie committed:
ingest-geoip: establish boundaries

This large-scale refactor is net zero change to the *usage* of the geoip
processor, but simplifies the public `GeoIpDatabase` interface to reduce
the complexity of implementing it from outside of the Elasticsearch codebase,
and to make implmentations less fragile as new database types are added
to this codebase.

Instead of needing to re-implement the interface each time we add support
for a new type, implementers will need to provide one final generic
implementation.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @yaauie, I've created a changelog YAML for you.

yaauie commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDatabase.java:
```diff
@@ -34,44 +30,18 @@ public interface GeoIpDatabase {
-
-    @Nullable
-    IspResponse getIsp(InetAddress ipAddress);
+    <RESPONSE extends AbstractResponse> RESPONSE getResponse(
```
The `GeoIpDatabase` interface is still tightly-coupled with maxmind's `DatabaseReader`, but it is no longer tightly-coupled with the _types_ of responses we can get from it, so implementations of this interface will no longer need to be modified when we add support for new types here.

yaauie replied:
There is an opportunity here to use [`<T> com.maxmind.db.Reader#get(InetAddress, Class<T>)`](https://maxmind.github.io/MaxMind-DB-Reader-java/doc/v3.1.0/com.maxmind.db/com/maxmind/db/Reader.html) _directly_ instead of proxying through to the model-bound `com.maxmind.geoip2.DatabaseReader.tryXXX(InetAddress)` methods, which would allow for the use of mmdb's whose contents do not adhere to one of the geoip2 models.

This would obviously be a larger refactor that would reach into the node service's Geoip provider that currently provides `com.maxmind.geoip2.DatabaseReader`, but may be required if we ever support mmdb's from other vendors.

yaauie commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoDataLookupFactory.java:
```diff
@@ -0,0 +1,29 @@
+interface GeoDataLookupFactory<PROVIDER extends GeoDataLookup> {
+    PROVIDER create(Set<Database.Property> properties);
+
+    static GeoDataLookupFactory<?> get(final Database database) {
```
`Database` still owns the mapping of which properties are available/default for each of these lookups. We could likely benefit from either consolidating them or making `Database` implement `GeoDataLookupFactory` instead of requiring this separate 1:1 lookup.

yaauie committed:
Update docs/changelog/109655.yaml

elasticsearchmachine commented:
Hi @yaauie, I've updated the changelog YAML for you.

robbavey commented:
@joegallo Would this be something you would be able to look at/start the conversation with, or point us towards someone else who might be able to?

This is something we would like to facilitate to help us deal with potential code changes in Elasticsearch causing unintended breaking changes in the Logstash integration filter.

Please reach out to me, @yaauie and/or @jsvd if you have any questions

joegallo committed:
Remove unused constants

joegallo committed:
Merge branch 'main' into ingest-geoip-boundary-refactor

yaauie committed:
Delete docs/changelog/109655.yaml

joegallo committed:
Handle nulls centrally

joegallo committed:
Mark these as final

joegallo committed:
Rename this method

joegallo committed:
Merge branch 'main' into ingest-geoip-boundary-refactor

joegallo committed:
Rename these variables

joegallo committed:
Strip off a layer of Optionals throughout

joegallo committed:
Rename these interfaces

joegallo committed:
Inline this method

joegallo committed:
Rename these local variables

joegallo committed:
Merge branch 'main' into ingest-geoip-boundary-refactor

joegallo committed:
Merge branch 'main' into ingest-geoip-boundary-refactor

joegallo committed:
Fix a typo

joegallo committed:
Tidy up the utility class -ing

joegallo committed:
This can be a link (rather than a code)

joegallo committed:
Whitespace

masseyke commented on modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/MaxMindSupportTests.java:
```diff
@@ -472,7 +492,7 @@ public void testUnknownMaxMindResponseClassess() {
      */
     public void testUsedMaxMindResponseClassesAreAccountedFor() {
-        Set<Class<? extends AbstractResponse>> usedMaxMindResponseClasses = getUsedMaxMindResponseClasses();
+        Set<Class<? extends AbstractResponse>> usedMaxMindResponseClasses = MAX_MIND_CLASS_TO_IMPLEMENTATION_CLASS.keySet();
```
This method is now pointless -- it's just comparing two sets that we define in the test. Ideally, we could get the set of maxmind responses that we use from IpDataLookupFactory. I wouldn't consider it a blocker for this PR, but if we don't do something like that we might as well delete this method so we don't get a false sense of security from it.

joegallo replied:
Fair, I dropped the test in a5d01f94025dd0fa4b8fa85bb26cb5dab311473b, and I think you and I should take as a task that it would be nice to do _something_ here (but not necessarily on this PR).

joegallo committed:
Merge branch 'main' into ingest-geoip-boundary-refactor

joegallo committed:
Emphasize IpDataLookup (and de-emphasize Database)

This moves a bunch of the logic from Database to
IpDataLookupFactories. There are still some default-visibility
internal details that are exposed for use in tests, but the primary
interface is the final 'get(String, String)' at the end of the file.

joegallo committed:
Merge branch 'main' into ingest-geoip-boundary-refactor

joegallo committed:
Drop this test until we can rewrite it

yaauie commented:
I did a first-pass implementing the new interfaces @joegallo added on the `logstash-filter-elastic_integration` side, and was able to get everything working aside from some of our plugin's internals being a little too aggressive at propagating Closeaeble#close, which will be easy enough to chase down. I have some other high-priority work, so I'll get back to an actual review of your PR shortly, but I'm pretty confident that the shape it exposes is good.

It would be _nice_ to be able to use `DatabaseReaderLazyLoader` and `GeoIPCache`, but they are not public. Without them we need to implement caching on our own or make our IpDatabase hold a Reader instantiated with the less-efficient cache.

masseyke committed:
merging main

masseyke committed:
making DatabaseReaderLazyLoader and GeoIpCache public

joegallo commented:
We might rewrite the cache and loader for 8.17 or 8.18, I have some ideas on the back burner. I don't mind them being public now, though, (and I see Keith made them public for you already, which I completely agree with) -- we'll just have to be in touch with you if we do end up making any changes.

I'd like to merge this PR, there's some additional work in the pipeline that builds on it.

@yaauie is there anything else you want to see here now or can we merge it any iterate as necessary?

joegallo committed:
Merge branch 'main' into ingest-geoip-boundary-refactor

joegallo commented:
Scratch that -- I think we're close enough here that this is mergeable, so I'm merging it. And if we need to iterate, then we can definitely still iterate (via more PRs!).

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 109655`
